### PR TITLE
GPU Profiling

### PR DESCRIPTION
**Pull request by BaronKiko**
This PR intends to implement profiling for the GPU.
To run this you must have the parent PR: https://github.com/Ryujinx/Ryujinx/pull/567
It was separated to make it easier to review, parent is already 2.4k-ish lines of code at time of writing

### CODE REVIEW

**Comment by EmulationChannel** [comment]
SMO and more GAMES 
sharedscreenshot2

**Comment by BaronKiko** [comment]
@EmulationChannel Have you got your fonts set up for ryujinx? It's using FontStandard.ttf in "RyuFs\system\fonts\"
If it's just that then I should probably make sure there is a clearer error message.
Also what platform are you on?

**Comment by jduncanator** [comment]
He’ll be using the official Nintendo fonts dumped from a Switch which have thousands more glyphs than a standard font. It may be that SharpFont doesn’t like so many glyphs, or the glyphs are using unsupported functionality of SharpFont.

**Comment by jduncanator** [comment]
If you just use Arial, it’s available everywhere.

**Comment by BaronKiko** [comment]
I would rather use a monospace font. Arial is a good fallback. My main concern is getting the location of the font across all platforms but Google will probably have an answer.
Anyway I will look into it when I get a moment and post the fix to the mail profiling branch. Till then you can just replace fontstandard with a normal font.

**Comment by jduncanator** [comment]
FontStandard.ttf fallback is Open Sans (I believe) which isn't monospace at the moment. If you're after a cross-platform Monospace font, use Courier.

**Comment by BaronKiko** [comment]
I know font standard isn't mono but decided to live with it as it was already required. Now I'm gonna change that I might as well use mono space. Courier will work well.

**Comment by BaronKiko** [comment]
Updated for merged profiler.
No longer WIP, ready for review.

File: Ryujinx.Profiler/ProfileConfig.cs [hunk 1]
```diff
@@ -105,6 +105,132 @@ public static class CPU
...
+                    Category     = "GPU.Engine2D",
+                    SessionGroup = "TextureCopy"
+                };
+            }
+
+            public static class Engine3d
+            {
+                public static ProfileConfig CallMethod = new ProfileConfig()
+                {
+                    Category = "GPU.Engine3D",
```
Comment (AcK77) [inline Ryujinx.Profiler/ProfileConfig.cs]:
```suggestion
                    Category     = "GPU.Engine3D",
```

File: Ryujinx.Graphics/Graphics3d/NvGpuEngineP2mf.cs [hunk 2]
```diff
@@ -54,7 +55,12 @@ public void CallMethod(NvGpuVmm vmm, GpuMethodCall methCall)
         {
             if (_methods.TryGetValue(methCall.Method, out NvGpuMethod method))
             {
+                ProfileConfig profile = Profiles.GPU.EngineP2mf.PushData;
+                profile.SessionItem   = method.Method.Name;
```
Comment (AcK77) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngineP2mf.cs]:
```suggestion

                profile.SessionItem = method.Method.Name;
```

File: Ryujinx.Graphics/Graphics3d/NvGpuEngineM2mf.cs [hunk 3]
```diff
@@ -37,7 +38,12 @@ public void CallMethod(NvGpuVmm vmm, GpuMethodCall methCall)
         {
             if (_methods.TryGetValue(methCall.Method, out NvGpuMethod method))
             {
+                ProfileConfig profile = Profiles.GPU.EngineM2mf.CallMethod;
+                profile.SessionItem   = method.Method.Name;
```
Comment (AcK77) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngineM2mf.cs]:
```suggestion

                profile.SessionItem = method.Method.Name;
```

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 4]
```diff
@@ -178,10 +195,16 @@ private void ClearBuffers(NvGpuVmm vmm, GpuMethodCall methCall)
...
             _gpu.Renderer.Pipeline.ResetDepthMask();
             _gpu.Renderer.Pipeline.ResetColorMask(attachment);
+
+            Profile.End(Profiles.GPU.Engine3d.ClearBuffers);
         }
 
         private void SetFrameBuffer(NvGpuVmm vmm, int fbIndex)
         {
+            ProfileConfig profile = Profiles.GPU.Engine3d.SetFrameBuffer;
+            profile.SessionItem   = fbIndex.ToString();
```
Comment (AcK77) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]:
```suggestion

            profile.SessionItem = fbIndex.ToString();
```

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 5]
```diff
@@ -178,10 +195,16 @@ private void ClearBuffers(NvGpuVmm vmm, GpuMethodCall methCall)
...
+            Profile.Begin(profile);
```
Comment (AcK77) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]:
```suggestion

            Profile.Begin(profile);
```

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 6]
```diff
@@ -89,7 +90,13 @@ public void CallMethod(NvGpuVmm vmm, GpuMethodCall methCall)
         {
             if (_methods.TryGetValue(methCall.Method, out NvGpuMethod method))
             {
+                ProfileConfig profile = Profiles.GPU.Engine3d.CallMethod;
+                profile.SessionItem = method.Method.Name;
```
Comment (AcK77) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]:
```suggestion

                profile.SessionItem = method.Method.Name;
```

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 7]
```diff
@@ -89,7 +90,13 @@ public void CallMethod(NvGpuVmm vmm, GpuMethodCall methCall)
...
+                Profile.Begin(profile);
```
Comment (AcK77) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]:
```suggestion

                Profile.Begin(profile);
```

**Comment by BaronKiko** [comment]
Applied Ac_K's suggestions. Should be good to go

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 8]
```diff
@@ -190,6 +217,7 @@ private void SetFrameBuffer(NvGpuVmm vmm, int fbIndex)
             {
                 _gpu.Renderer.RenderTarget.UnbindColor(fbIndex);
 
+                Profile.End(profile);
```
Comment (AcK77) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]:
```suggestion
                Profile.End(profile);

```

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 9]
```diff
@@ -264,6 +296,7 @@ private void SetZeta(NvGpuVmm vmm)
             {
                 _gpu.Renderer.RenderTarget.UnbindZeta();
 
+                Profile.End(Profiles.GPU.Engine3d.SetZeta);
```
Comment (AcK77) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]:
```suggestion
                Profile.End(Profiles.GPU.Engine3d.SetZeta);

```

**Comment by BaronKiko** [comment]
Applied suggested changes

**Pull request merged**